ql/src/java/org/apache/hadoop/hive/ql/Driver.java

```diff
@@ -588,15 +353,280 @@ private void runInternal(String command, boolean alreadyCompiled) throws Command
+        driverTxnHandler.cleanupTxnList();
+      }
+    }
+  }
+
+  private void compileInternal(String command, boolean deferClose) throws CommandProcessorException {
+    Metrics metrics = MetricsFactory.getInstance();
+    if (metrics != null) {
+      metrics.incrementCounter(MetricsConstant.WAITING_COMPILE_OPS, 1);
+    }
+
+    PerfLogger perfLogger = SessionState.getPerfLogger(true);
+    perfLogger.perfLogBegin(CLASS_NAME, PerfLogger.WAIT_COMPILE);
+
+    try (CompileLock compileLock = CompileLockFactory.newInstance(driverContext.getConf(), command)) {
+      boolean success = compileLock.tryAcquire();
+
+      perfLogger.perfLogEnd(CLASS_NAME, PerfLogger.WAIT_COMPILE);
+
+      if (metrics != null) {
+        metrics.decrementCounter(MetricsConstant.WAITING_COMPILE_OPS, 1);
+      }
+      if (!success) {
+        String errorMessage = ErrorMsg.COMPILE_LOCK_TIMED_OUT.getErrorCodedMsg();
+        throw DriverUtils.createProcessorException(driverContext, ErrorMsg.COMPILE_LOCK_TIMED_OUT.getErrorCode(),
+            errorMessage, null, null);
+      }
+
+      try {
+        compile(command, true, deferClose);
+      } catch (CommandProcessorException cpe) {
+        try {
+          driverTxnHandler.endTransactionAndCleanup(false);
+        } catch (LockException e) {
+          LOG.warn("Exception in releasing locks. " + StringUtils.stringifyException(e));
+        }
+        throw cpe;
+      }
+    }
+    //Save compile-time PerfLogging for WebUI.
+    //Execution-time Perf logs are done by either another thread's PerfLogger
+    //or a reset PerfLogger.
+    driverContext.getQueryDisplay().setPerfLogStarts(QueryDisplay.Phase.COMPILATION, perfLogger.getStartTimes());
+    driverContext.getQueryDisplay().setPerfLogEnds(QueryDisplay.Phase.COMPILATION, perfLogger.getEndTimes());
+  }
+
+  /**
+   * Compiles a new HQL command, but potentially resets taskID counter. Not resetting task counter is useful for
+   * generating re-entrant QL queries.
+   *
+   * @param command  The HiveQL query to compile
+   * @param resetTaskIds Resets taskID counter if true.
+   * @return 0 for ok
+   */
+  public int compile(String command, boolean resetTaskIds) {
+    try {
+      compile(command, resetTaskIds, false);
+      return 0;
+    } catch (CommandProcessorException cpr) {
+      return cpr.getErrorCode();
+    }
+  }
+
+  /**
+   * Compiles an HQL command, creates an execution plan for it.
+   *
+   * @param deferClose indicates if the close/destroy should be deferred when the process has been interrupted, it
+   *        should be set to true if the compile is called within another method like runInternal, which defers the
+   *        close to the called in that method.
+   * @param resetTaskIds Resets taskID counter if true.
+   */
+  @VisibleForTesting
+  public void compile(String command, boolean resetTaskIds, boolean deferClose) throws CommandProcessorException {
```

Comment:
Please change order of `@param` to match arguments and also add one for 'command' argument.